modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modeling
+
+import (
+	"fmt"
+)
+
+// LocalSliceAlgorithm 'borrows' and 'rents' endpoints from other zones to make
+// the local EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution):
+// 1. The proportion of incoming traffic for a zone is calculated based on the
+// proportion of nodes or cores in the zone.
+// 2. This is compared with the proportion of endpoints in the zone to calculate
+// the deviation from the expected number of endpoints in a perfectly balanced
+// system.
+// 3. EndpointSlices in zones with less endpoints than expected will receive
+// endpoints from zones that have more endpoints than expected.
+type LocalSliceAlgorithm struct{}
+
+type endpointDeviation struct {
+	name      string
+	deviation int
+}
+
+// endpointsList maintains the order of the endpoints available/needed, used as
+// a list
+type endpointsList struct {
+	byZone []endpointDeviation
+}
+
+// push to back
+func (el *endpointsList) push(newValue endpointDeviation) {
+	el.byZone = append(el.byZone, newValue)
+}
+
+// pop the front
+func (el *endpointsList) pop() {
+	if len(el.byZone) == 0 {
+		return
+	}
+	el.byZone = el.byZone[1:]
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// endpointsAvailable stores 'endpointDeviation' -- zones with 'extra'
+	// endpoints
```

Comment:
Nit: Not sure you need to mention `endpointDeviation` here - I think the comment here could mirror the pattern you used below for `endpointsNeeded`.

Reply:
changed